modules/core/src/smithy4s/Service.scala

```diff
@@ -187,4 +187,84 @@ object Service {
     final def toPolyFunction[P[_, _, _, _, _]](algebra: PolyFunction5[Op, P]): PolyFunction5[Op, P] = algebra
     final def mapK5[F[_, _, _, _, _], G[_, _, _, _, _]](algebra: PolyFunction5[Op, F], function: PolyFunction5[F, G]): PolyFunction5[Op, G] = algebra.andThen(function)
   }
+
+  object Builder {
+    def fromService[Alg[_[_, _, _, _, _]]](
+        service: Service[Alg]
+    ): Builder[Alg, service.Operation] =
+      new Builder[Alg, service.Operation](service, PolyFunction5.identity, service.id, service.version, service.hints)
+  }
+
+  final case class Builder[Alg[_[_, _, _, _, _]], Op[_, _, _, _, _]] private(
+      private val base: Service.Aux[Alg, Op],
+      private val endpointMapper: PolyFunction5[Endpoint.ForOperation[Op]#e, Endpoint.ForOperation[Op]#e],
+      private val baseId: ShapeId,
+      private val baseVersion: String,
+      private val baseHints: Hints,
+  ) {
+
+    def mapEndpointEach(
+        mapper: PolyFunction5[Endpoint.ForOperation[Op]#e, Endpoint.ForOperation[Op]#e]
+    ): Builder[Alg, Op] = copy(
+      endpointMapper = (endpointMapper.andThen(mapper)).unsafeCacheBy(
```

Comment:
**thought:** this transformation is kind of lazy (only happens when you call `.build`), perhaps we should document that in its Scaladoc.

also, perhaps we should only cache it inside `build`, otherwise if you call `mapEndpointEach` multiple times you get multiple layers of caching.

Reply:
NB : syncing with series `0.18`  will remove `unsafeCacheBy`, but it's okay because the exposure of ordinals should remove the need to cache.

Reply:
it did, because we got rid of `def endpoint`. We'll defer to the default impl, unless you prefer otherwise.